### PR TITLE
New lint: detect unnecessary struct building

### LINKED ISSUE #10476: Warn on redundant update
### What it does

If your update changes no fields, e.g. `MyRecord {..value}` then the update is providing very little value, and `value` directly would be better.

### Lint Name

redundant_update

### Category

style

### Advantage

- Simpler.

### Drawbacks

_No response_

### Example

```rust
MyRecord { ..value}
```

Could be written as:

```rust
value
```

### PR DESCRIPTION
**Pull request by samueltardieu**
Fixes #10476.

Running this lint on the top 500 crates produced one hit (in `rust-lang/rust-bindgen`) and [a PR has been submitted there](https://github.com/rust-lang/rust-bindgen/pull/2440).

changelog: [`unnecessary_struct_initialization`]: new lint

### CODE REVIEW

**Comment by samueltardieu** [comment]
Withdrawn, I hadn't noticed `needless_update`, I will submit a patch to make it more efficient instead.

**Comment by giraffate** [comment]
I'll review this in this week.

File: clippy_lints/src/unnecessary_struct.rs [hunk 1]
```diff
@@ -0,0 +1,84 @@
...
+    /// ```
+    /// Use instead:
+    /// ```rust
+    /// struct S { s: String }
+    ///
+    /// let a = S { s: String::from("Hello, world!") };
+    /// let b = a;
+    /// ```
+    #[clippy::version = "1.70.0"]
+    pub UNNECESSARY_STRUCT,
```
[the remark this answers is not in the document]
Comment (samueltardieu) [reply clippy_lints/src/unnecessary_struct.rs]: Good idea. Fixed.

**Comment by giraffate** [comment]
@bors r+

Thanks!

**Pull request merged**